### PR TITLE
Integrate with Service Connect AWSVPC CNI plugin

### PR DESCRIPTION
**Pull request by yinyic**
### Summary

ServiceConnect-enabled tasks require network namespace configurations that are performed by a new CNI plugin. This PR integrates with the new CNI plugin `ecs-serviceconnect` for AWSVPC mode. 

### Implementation details

If a task is AWSVPC mode and SC-enabled, build and append ServiceConnect network configuration such that the `ecs-serviceconnect` CNI plugin will be invoked with the supplied information. 

The configuration includes:
- ingress and egress info which comes from task payload (except the ephemeral ports which are selected by Agent)
- flags to indicate whether task is IPv4- and/or IPv6-enabled

The PR also upgrades the submodule commit for `amazon-vpc-cni-plugins` to pick up the new SC plugin.

### Testing

New tests cover the changes:  yes

Built test Agent and ran a test SC task - verified netns configuration for both default and non-default SC experience in AWSVPC network mode.

For CNI plugin build, ran `make cni-plugins` for both `TARGET_OS=linux` and `TARGET_OS=windows` - verified all CNI plugins were built successfully.

### Description for the changelog

Integrate with Service Connect AWSVPC CNI plugin

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/task/task_linux.go
Comment (rajal-amzn) [inline agent/api/task/task_linux.go:353]: Which interface is set here?

File: agent/ecscni/netconfig_linux.go
Comment (rajal-amzn) [inline agent/ecscni/netconfig_linux.go:184]: Why cant we simply copy the Ingress Config entry?

File: agent/ecscni/netconfig_linux.go
Comment (rajal-amzn) [inline agent/ecscni/netconfig_linux.go:202]: We also need to add `NetworkMode`

File: agent/ecscni/types_linux.go
Comment (rajal-amzn) [inline agent/ecscni/types_linux.go:33]: Why cant we actually send the interface name available to this container? The interface name is not used in awsvpc but it would be for bridge.

File: agent/api/task/task_linux.go
Comment (yinyic) [reply agent/api/task/task_linux.go:353]: Resolving as dup of the [this comment](https://github.com/aws/amazon-ecs-agent/pull/3255#discussion_r899356403)

File: agent/ecscni/netconfig_linux.go
Comment (yinyic) [reply agent/ecscni/netconfig_linux.go:184]: Not sure I follow - the `IngressConfigJSONEntry` is a different struct from `scConfig.IngressConfig` entries. Former is the internal model for CNI while latter is the SC API model.

File: agent/ecscni/netconfig_linux.go
Comment (yinyic) [reply agent/ecscni/netconfig_linux.go:202]: Do you mean `RedirectMode`? Will do once the CNI plugin PR is merged. The CNI version contained in this PR does not have that `RedirectMode` yet.

File: agent/ecscni/types_linux.go
Comment (yinyic) [reply agent/ecscni/types_linux.go:33]: I saw the CNI plugin CR using `eth0` (which indeed is the default ifname for docker bridge netns) so I didn't add anything here. I can add it.

File: agent/ecscni/types_linux.go
Comment (yinyic) [reply agent/ecscni/types_linux.go:168]: I believe this is the recommended way actually, according to https://go.dev/doc/effective_go#package-names
> The importer of a package will use the name to refer to its contents, so exported names in the package can use that fact to avoid repetition

In our case, the package name `ecscni` already implies the namespace, and other packages will be referencing the type as `ecscni.ServiceConnectConfig`.

File: agent/ecscni/types_linux.go
Comment (rajal-amzn) [reply agent/ecscni/types_linux.go:33]: Yes I changed it to use `ifName` in the latest revision.

File: agent/ecscni/netconfig_linux.go
Comment (chienhanlin) [inline agent/ecscni/netconfig_linux.go:212]: Borrow wisdom from Angel ([source](https://github.com/aws/amazon-ecs-agent/pull/3208#discussion_r870645566))
>nit: the preferred way to wrap errors is now fmt.Errorf("message: %w", err)
github.com/pkg/errors is a bit of a legacy package now

File: agent/ecscni/plugin_linux_test.go
Comment (chienhanlin) [inline agent/ecscni/plugin_linux_test.go:290]: nit. maybe we can define 
```
const (
  testIngressListenerPort = uint16(11111)
  testEgressConfigListenerPort = uint16(22222)
  ....
)
```
and reuse then in test validation steps.

File: Makefile
Comment (yinyic) [reply Makefile:212]: We need network to download the dependencies. 

According to https://go.dev/ref/mod#mod-commands

> In module-aware mode, the go command uses go.mod files to find versioned dependencies, and it typically loads packages out of the [module cache](https://go.dev/ref/mod#glos-module-cache), downloading modules if they are missing. 

We used to use vendor directory before that's why it does not require network. 

> In GOPATH mode, the go command ignores modules; it looks in [vendor directories](https://go.dev/ref/mod#glos-vendor-directory) and in GOPATH to find dependencies.

I'm making the change because the source amazon-vpc-cni-plugins [repo](https://github.com/aws/amazon-vpc-cni-plugins) also builds in module-aware mode. Although it has a vendor directory, the dependency version has been out-of-sync. 

If we believe the need for internet during a build should be avoided as much as possible, we can go back to using vendor directory by asking the CNI plugin maintainers to keep the vendor directory up to date.

File: agent/ecscni/types_linux.go
Comment (yinyic) [reply agent/ecscni/types_linux.go:33]: Cool, will address in next revision.

File: agent/ecscni/netconfig_linux.go
Comment (yinyic) [reply agent/ecscni/netconfig_linux.go:212]: Sure will address in next rev

File: agent/ecscni/netconfig_linux.go
Comment (yinyic) [reply agent/ecscni/netconfig_linux.go:210]: Sure will update.

File: agent/ecscni/plugin_linux_test.go
Comment (yinyic) [reply agent/ecscni/plugin_linux_test.go]: Yeah sure!

File: agent/ecscni/netconfig.go
Comment (chienhanlin) [inline agent/ecscni/netconfig.go:29]: Thanks!

**Review by chienhanlin — DISMISSED** [review]

File: Makefile
Comment (rajal-amzn) [reply Makefile:212]: I would suggest we go and fix CNI plugin to keep the vendor directory up to date.

File: agent/ecscni/netconfig_linux.go
Comment (rajal-amzn) [reply agent/ecscni/netconfig_linux.go:184]: I was suggesting that we could directly use the scConfig.IngressConfig given it only has some additional fields and they would be ignored anywhere. But given that the number of entries in here will be less. it might not matter.

Comment (yinyic) [reply agent/ecscni/netconfig_linux.go:184]: Yeah I think I'd rather keep it separate. The EgressConfig that we pass to CNI plugin will have to be different than the API model anyway (to include `RedirectMode` for example), so it's probably better to keep IngressConfig separate as well.

File: Makefile
Comment (yinyic) [reply Makefile:212]: Chatted with @ofiliz, maintainer of `amazon-vpc-cni-plugins` repo - Onur is going to handpick all necessary dependencies because running `go mod vendor` brings a much larger dependency set than what we actually need and is undesirable.

Comment (yinyic) [reply Makefile:212]: https://github.com/aws/amazon-vpc-cni-plugins/pull/73

**Review by aws-gibbskt — APPROVED** [review]

File: agent/ecscni/types_linux.go
Comment (ofiliz) [inline agent/ecscni/types_linux.go:36]: I don't think this is accurate. There should be no interface names hardcoded in CNI plugins or agent. We have removed the use of ifName from ServiceConnect CNI plugin.

Comment (rajal-amzn) [reply agent/ecscni/types_linux.go:36]: That is right. We dont use the ifName within the plugin.

Comment (yinyic) [reply agent/ecscni/types_linux.go:36]: Okay will update.

File: agent/ecscni/plugin_test.go
Comment (aws-gibbskt) [inline agent/ecscni/plugin_test.go:33]: 24d6bd87707d1b1801086fc507ebab8d32067412 this is the latest after `Adding a new mode Tproxy to do the redirection of traffic for ServiceConnect Plugin (#74)`

Comment (yinyic) [reply agent/ecscni/plugin_test.go:33]: The latest CNI plugin repo commit updated request model which is incompatible with the code in this PR. In my follow-up PR for bridge mode, both the model and the CNI git hash will be updated. I wanted to keep PR to be a simple rebase if possible.

**Review by rajal-amzn — APPROVED** [review]

**Review by aws-gibbskt — APPROVED** [review]

**Review by chienhanlin — APPROVED** [review]

**Review by ofiliz — APPROVED** [review]

**Review by ubhattacharjya — APPROVED** [review]

**Pull request merged**
